src/pybind/mgr/volumes/fs/volume.py

```diff
@@ -11,9 +13,147 @@
 
 log = logging.getLogger(__name__)
 
+class ConnectionPool(object):
+    class Connection(object):
+        def __init__(self, mgr, fs_name):
+            self.fs = None
+            self.mgr = mgr
+            self.fs_name = fs_name
+            self.ops_in_progress = 0
+            self.last_used = time.time()
+            self.fs_id = self.get_fs_id()
+
+        def get_fs_id(self):
+            fs_map = self.mgr.get('fs_map')
+            for fs in fs_map['filesystems']:
+                if fs['mdsmap']['fs_name'] == self.fs_name:
+                    return fs['id']
+            raise VolumeException(
+                -errno.ENOENT, "Volume '{0}' not found".format(self.fs_name))
+
+        def get_fs_handle(self):
+            self.last_used = time.time()
+            self.ops_in_progress += 1
+            return self.fs
+
+        def put_fs_handle(self):
+            assert self.ops_in_progress > 0
+            self.ops_in_progress -= 1
+
+        def del_fs_handle(self):
+            if self.is_connection_valid():
+                self.disconnect()
+            else:
+                self.abort()
+
+        def is_connection_valid(self):
+            fs_id = None
+            try:
+                fs_id = self.get_fs_id()
+            except:
+                # the filesystem does not exist now -- connection is not valid.
+                pass
+            return self.fs_id == fs_id
+
+        def is_connection_idle(self, timeout):
+            return (self.ops_in_progress == 0 and ((time.time() - self.last_used) >= timeout))
+
+        def connect(self):
+            assert self.ops_in_progress == 0
+            log.debug("Connecting to cephfs '{0}'".format(self.fs_name))
+            self.fs = cephfs.LibCephFS(rados_inst=self.mgr.rados)
+            log.debug("CephFS initializing...")
+            self.fs.init()
+            log.debug("CephFS mounting...")
+            self.fs.mount(filesystem_name=self.fs_name.encode('utf-8'))
+            log.debug("Connection to cephfs '{0}' complete".format(self.fs_name))
+
+        def disconnect(self):
+            assert self.ops_in_progress == 0
+            log.info("disconnecting from cephfs '{0}'".format(self.fs_name))
+            self.fs.shutdown()
+            self.fs = None
+
+        def abort(self):
+            assert self.ops_in_progress == 0
```

Comment:
I hit this assertion, when I ran the reran the test locally on my laptop (after changing the test to purge lesser number of files),

```
2019-06-21 22:34:48,816.816 INFO:__main__:Running ['./bin/ceph', '--admin-daemon', '/tmp/ceph-asok.0G44bX//client.0.86970.asok', 'status']
2019-06-21 22:34:48,937.937 INFO:__main__:Running ['stat', '--file-system', '--printf=%T\n', '--', '/tmp/tmpWnZF2z/mnt.0']
2019-06-21 22:34:48,950.950 INFO:tasks.cephfs.fuse_mount:ceph-fuse is mounted on /tmp/tmpWnZF2z/mnt.0
2019-06-21 22:34:48,950.950 INFO:__main__:Running ['chmod', '1777', '/tmp/tmpWnZF2z/mnt.0']
2019-06-21 22:34:48,964.964 INFO:__main__:Running ['./bin/ceph', 'fs', 'flag', 'set', 'enable_multiple', 'true', '--yes-i-really-mean-it']
2019-06-21 22:34:50,227.227 INFO:__main__:Running ['./bin/ceph', 'fs', 'volume', 'ls']
2019-06-21 22:34:50,568.568 INFO:__main__:Running ['./bin/ceph', 'fs', 'subvolume', 'create', u'cephfs', 'subvolume_8387']
2019-06-21 22:34:50.690 7fe766a06700 -1 WARNING: all dangerous and experimental features are enabled.
2019-06-21 22:34:50.754 7fe766a06700 -1 WARNING: all dangerous and experimental features are enabled.
Error EINVAL: Traceback (most recent call last):
  File "/home/rraja/git/ceph/src/pybind/mgr/mgr_module.py", line 889, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/module.py", line 177, in handle_command
    return handler(inbuf, cmd)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/module.py", line 216, in _cmd_fs_subvolume_create
    size=cmd.get('size', None))
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/volume.py", line 341, in conn_wrapper
    fs_h = self.connection_pool.get_fs_handle(fs_name)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/volume.py", line 128, in get_fs_handle
    self._del_fs_handle(fs_name)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/volume.py", line 149, in _del_fs_handle
    conn.del_fs_handle()
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/volume.py", line 48, in del_fs_handle
    self.abort()
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/volume.py", line 79, in abort
    assert self.ops_in_progress == 0
AssertionError

2019-06-21 22:34:50,914.914 INFO:__main__:test_async_subvolume_rm (tasks.cephfs.test_volumes.TestVolumes) ... ERROR
2019-06-21 22:34:50,914.914 ERROR:__main__:Traceback (most recent call last):
  File "/home/rraja/git/ceph/qa/tasks/cephfs/test_volumes.py", line 369, in test_async_subvolume_rm
    self._fs_cmd("subvolume", "create", self.volname, subvolume)
  File "/home/rraja/git/ceph/qa/tasks/cephfs/test_volumes.py", line 27, in _fs_cmd
    return self.mgr_cluster.mon_manager.raw_cluster_cmd("fs", *args)
  File "../qa/tasks/vstart_runner.py", line 696, in raw_cluster_cmd
    proc = self.controller.run([os.path.join(BIN_PREFIX, "ceph")] + list(args), **kwargs)
  File "../qa/tasks/vstart_runner.py", line 342, in run
    proc.wait()
  File "../qa/tasks/vstart_runner.py", line 179, in wait
    raise CommandFailedError(self.args, self.exitstatus)
CommandFailedError: Command failed with status 22: ['./bin/ceph', 'fs', 'subvolume', 'create', u'cephfs', 'subvolume_8387']

2019-06-21 22:34:50,914.914 ERROR:__main__:Error in test 'test_async_subvolume_rm (tasks.cephfs.test_volumes.TestVolumes)', going interactive
Ceph test interactive mode, use ctx to interact with the cluster, press control-D to exit...
```

Reply:
> I hit this assertion, when I ran the reran the test locally on my laptop (after changing the test to purge lesser number of files),

this hints me about a timing issue although this test passes with around 10 odd files. do you have a reproducer? something like running a particular test in a loop?

the backtrace tells me that the filesystem was removed and recreated in between tests, and that's fine as we handle the case. plus the purge thread operations go through the connection pool and the test waits for the threads to finish the purge.

Reply:
in short -- the filesystem got removed (and recreated) when an operation was in progress.

Reply:
thinking more on this -- `abort()` is called when a given filesystem handle in the connection pool is stale (i.e., the filesystem got removed without us knowing). At this point, it is possible that file system operations are in-progress. The assert here is put under the assumption that there would not be any in-progress operations when cleaning up a stale handle (for say, teuthology), which I think might not be really true generally. We can remove this assertion and test that during abort pending filesystem operations return with an errcode (-ESTALE maybe?) and that should be good enough.

I would really like a reproducer test for this failure.